File: shoring/flywheel/src/main/java/com/panopset/flywheel/SourceFile.java
package com.panopset.flywheel;

import java.io.File;
import java.util.List;
import com.panopset.compat.Fileop;

/**
 * Flywheel source file, all paths relative to the primary script file.
 */
public final class SourceFile {

  private final String relativePath;

  String getRelativePath() {
    return relativePath;
  }

  private final File file;

  File getFile() {
    return file;
  }

  private final String canonicalPath;

  private List<String> sourceLines;

  /**
   * Directory depth relative to driver template.
   */
  private Integer depth;

  /**
   * Get depth relative to driver template.
   *
   * @return Directory depth relative to driver template
   */
  public int getDepth() {
    if (depth == null) {
      depth = 0;
      String temp = relativePath;
      int idx = temp.indexOf("/");
      while (idx > -1) {
        depth++;
        temp = temp.substring(idx + 1);
        idx = temp.indexOf("/");
      }
    }
    return depth;
  }

  /**
   * Reset source lines, for list processing.
   */
  public void reset() {
    sourceLines = null;
  }

  /**
   * Get source lines.
   *
   * @return Source lines.
   */
  public List<String> getSourceLines() {
    if (sourceLines == null) {
      sourceLines = Fileop.readLines(file);
    }
    return sourceLines;
  }

  /**
   * SourceFile constructor.
   *
   * @param sourceFile Source file.
   * @param basePath Base directory path.
   */
  public SourceFile(final File sourceFile, final String basePath) {
    file = sourceFile;
    canonicalPath = Fileop.getCanonicalPath(this.file);
    relativePath = canonicalPath.substring(basePath.length() + 1);
  }

  /**
   * SourceFile constructor.
   *
   * @param sourceFileRelativePath Source file relative path.
   * @param flywheel Flywheel.
   */
  public SourceFile(final Flywheel flywheel, final String sourceFileRelativePath) {
    relativePath = sourceFileRelativePath;
    file = new File(flywheel.getBaseDirectoryPath() + "/" + relativePath);
    canonicalPath = Fileop.getCanonicalPath(file);
  }

  /**
   * Check to see if source file exists, and can be read.
   *
   * @return true if source file exists.
   */
  public boolean isValid() {
    return file.exists() && file.canRead();
  }

  /**
   * Override equals method. From
   * <a href="http://www.geocities.com/technofundo/tech/java/equalhash.html">
   * http://www.geocities.com/technofundo/tech/java/equalhash.html </a>
   *
   * @param obj Object.
   * @return true if obj is equal to this.
   */
  @Override
  public boolean equals(final Object obj) {
    if (this == obj) {
      return true;
    }
    if ((obj == null) || (obj.getClass() != this.getClass())) {
      return false;
    }
    SourceFile test = (SourceFile) obj;
    return canonicalPath.equals(test.canonicalPath) && relativePath.equals(test.relativePath);
  }

  /**
   * @return hash code for this object.
   */
  @Override
  public int hashCode() {
    return file.hashCode();
  }
}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/flywheel/SourceFile.kt
package com.panopset.flywheel

import com.panopset.compat.Fileop.getCanonicalPath
import com.panopset.compat.Fileop.readLines
import java.io.File

/**
 * Flywheel source file, all paths relative to the primary script file.
 */
class SourceFile() {
    lateinit var relativePath: String
    lateinit var file: File
    lateinit var canonicalPath: String
    private var sourceLines: List<String> = ArrayList()

    /**
     * Directory depth relative to driver template.
     */
    var depth: Int? = null
        /**
         * Get depth relative to driver template.
         *
         * @return Directory depth relative to driver template
         */
        get() {
            if (field == null) {
                field = 0
                var temp = relativePath
                var idx = temp.indexOf("/")
                while (idx > -1) {
                    field = field!! + 1
                    temp = temp.substring(idx + 1)
                    idx = temp.indexOf("/")
                }
            }
            return field
        }
        private set

    /**
     * Reset source lines, for list processing.
     */
    fun reset() {
        sourceLines = ArrayList()
    }

    /**
     * Get source lines.
     *
     * @return Source lines.
     */
    fun getTheSourceLines(): List<String> {
        if (sourceLines.isEmpty()) {
            sourceLines = readLines( file)
        }
        return sourceLines
    }

    /**
     * SourceFile constructor.
     *
     * @param sourceFile Source file.
     * @param basePath Base directory path.
     */
    constructor(sourceFile: File, basePath: String): this() {
        file = sourceFile
        canonicalPath = getCanonicalPath(file)
        relativePath = canonicalPath.substring(basePath.length + 1)
    }

    /**
     * SourceFile constructor.
     *
     * @param sourceFileRelativePath Source file relative path.
     * @param flywheel Flywheel.
     */
    constructor(flywheel: Flywheel, sourceFileRelativePath: String): this() {
        relativePath = sourceFileRelativePath
        file = File(flywheel.getBaseDirectoryPath() + "/" + relativePath)
        canonicalPath = getCanonicalPath(file)
    }

    val isValid: Boolean
        /**
         * Check to see if source file exists, and can be read.
         *
         * @return true if source file exists.
         */
        get() = file.exists() && file.canRead()

    /**
     * Override equals method. From
     * [
 * http://www.geocities.com/technofundo/tech/java/equalhash.html ](http://www.geocities.com/technofundo/tech/java/equalhash.html)
     *
     * @param other Object.
     * @return true if obj is equal to this.
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || other.javaClass != this.javaClass) {
            return false
        }
        val test = other as SourceFile
        return canonicalPath == test.canonicalPath && relativePath == test.relativePath
    }

    /**
     * @return hash code for this object.
     */
    override fun hashCode(): Int {
        return file.hashCode()
    }
}
